iceoryx_utils/include/iceoryx_utils/internal/cxx/forward_list.inl

```diff
@@ -0,0 +1,538 @@
+{
+    if (this != &rhs)
+    {
+        uint64_t i = 0u;
+        auto iterThis = before_begin();
+        auto iterRhs = rhs.before_begin();
+        auto startSize = size();
+
+        // move using move assignment
+        for (; i < std::min(rhs.size(), size()); ++i)
+        {
+            ++iterThis;
+            ++iterRhs;
+            *iterThis = std::move(*iterRhs);
+        }
+
+        // move using move ctor (move additional elements from rhs.size to this.size)
+        for (; i < rhs.size(); ++i)
+        {
+            ++iterRhs;
+            iterThis = emplace_after(iterThis, std::move(*iterRhs));
+        }
+
+        // delete remaining elements
+        for (; i < startSize; ++i)
+        {
+            erase_after(iterThis);
+        }
+
+        rhs.clear();
+    }
+    return *this;
+}
+
+
+template <typename T, int64_t CAPACITY>
+forward_list<T, CAPACITY>::~forward_list()
+{
+    clear();
+}
+
+
+template <typename T, int64_t CAPACITY>
+typename forward_list<T, CAPACITY>::iterator forward_list<T, CAPACITY>::before_begin() noexcept
+{
+    return iterator{this, BEFORE_BEGIN_USED_INDEX};
+}
+template <typename T, int64_t CAPACITY>
+typename forward_list<T, CAPACITY>::const_iterator forward_list<T, CAPACITY>::before_begin() const noexcept
+{
+    return cbefore_begin();
+}
+template <typename T, int64_t CAPACITY>
+typename forward_list<T, CAPACITY>::const_iterator forward_list<T, CAPACITY>::cbefore_begin() const noexcept
+{
+    return const_iterator{this, BEFORE_BEGIN_USED_INDEX};
+}
+
+template <typename T, int64_t CAPACITY>
+typename forward_list<T, CAPACITY>::iterator forward_list<T, CAPACITY>::begin() noexcept
+{
+    iterator iter = before_begin();
+    return ++iter;
+}
+template <typename T, int64_t CAPACITY>
+typename forward_list<T, CAPACITY>::const_iterator forward_list<T, CAPACITY>::begin() const noexcept
+{
+    return cbegin();
+}
+template <typename T, int64_t CAPACITY>
+typename forward_list<T, CAPACITY>::const_iterator forward_list<T, CAPACITY>::cbegin() const noexcept
+{
+    const_iterator citer = cbefore_begin();
+    return ++citer;
+}
+
+
+template <typename T, int64_t CAPACITY>
+typename forward_list<T, CAPACITY>::iterator forward_list<T, CAPACITY>::end() noexcept
+{
+    return iterator{this, INVALID_INDEX};
+}
+template <typename T, int64_t CAPACITY>
+typename forward_list<T, CAPACITY>::const_iterator forward_list<T, CAPACITY>::end() const noexcept
+{
+    return cend();
+}
+template <typename T, int64_t CAPACITY>
+typename forward_list<T, CAPACITY>::const_iterator forward_list<T, CAPACITY>::cend() const noexcept
+{
+    return const_iterator{this, INVALID_INDEX};
+}
+
+
+template <typename T, int64_t CAPACITY>
+inline bool forward_list<T, CAPACITY>::empty() const noexcept
+{
+    return (m_size == 0);
+}
+
+template <typename T, int64_t CAPACITY>
+inline bool forward_list<T, CAPACITY>::full() const noexcept
+{
+    return (m_size >= CAPACITY);
```

Comment:
Can it ever happen that `m_size > CAPACITY`? If not could you please write it more explicitly `m_size == CAPACITY`

Reply:
I remember discussing about some unaddressed concurrency scenario which might end up 'm_size > CAPACITY'. We also check the iterator value on '>=INVALID_INDEX'. I guess it would be more consistent in that sense.